Step2 리뷰 요청

hyeongwon-up opened the pull request.

hyeongwon-up committed:
docs(README.md): 2단계 실습 요구사항 작성

hyeongwon-up committed:
feat(StringAddCalculatorTest.java): TestCase 소스코드 추가

hyeongwon-up committed:
feat(StringAddCalculator.java, StringAddCalculatorTest.java): StringCalculator.java 추가

- splitAndSum 메서드 추가.
- parseInput 메서드 추가.
- validateInput 메서드 추가.
- sum 메서드 추가.
- validateNegativeNumber 메서드 추가

hyeongwon-up committed:
feat: Split 과 StringSum 에 대한 기능 분리.

- feat(Split.java, StringSum.java): Split 클래스 생성. StringSum 클래스 생성.
- test(SplitTest.java): 정상적인 Split 동작을 위한 테스트.
- test(StringSum.java): 정상적인 String 배열의 합 동작 테스트.

hyeongwon-up committed:
refactor(Split.java, SplitTest.java): 필드 변수명 변경 value -> stringArray

hyeongwon-up committed:
refactor(StringSum.java): 필요없는 필드 제거.

hyeongwon-up committed:
refactor(StringAddCalculator.java): Split 과 StringSum 으로 분리된 기능 적용

geunhwanlee commented on src/main/java/calculator/READM.md:
```diff
@@ -0,0 +1,16 @@
+# 2단계 실습 - 문자열 덧셈 계산기
```
기능 요구사항 정리해주신것 좋습니다.👍🏻
다만 루트 디렉토리의 README.md 파일에 작성해주셔도 괜찮을것 같아요!

hyeongwon-up replied:
다음 3단계 로또 실습을 진행을 위한 기능 요구사항을 정리한 것도 루트에 같이 작성을 하는게 괜찮다는 말씀일까요?
각 단계에 대한 요구사항을 구분을 두어 정리하기 위해 해당 위치해 두었거든요!
이 부분에 대해서 답변 부탁 드리겠습니다!

geunhwanlee replied:
그런 이유에서라면 따로 작성하셔도 괜찮습니다!👍

이번 미션에서는 두가지의 과제를 하겠지만
다음 미션부터는 하나의 큰 과제를 진행하게 될거에요.
그때 루트디렉토리에 작성해주셔도 괜찮습니다.😄

geunhwanlee commented on src/main/java/calculator/Split.java:
```diff
@@ -0,0 +1,43 @@
+  public String[] validateInput(String input) {
+
+    if(input == null || input.isEmpty()) {
+      return new String[] {NULL_INPUT_VALUE};
```
문자열을 구분자로 분리하는 객체인 Split에서
input이 null 또는 공백일때 0이 들어있는 단일배열을 리턴하는건 부자연스럽네요.🤔
input이 null 또는 공백일때 빈 배열을 리턴하는게 더 자연스럽지 않을까 생각합니다.

hyeongwon-up replied:
다시 생각해보니 리뷰어님 의견에 동의합니다.
Split 에서 0 을 리턴하는건 어색하다고 생각이들고, 해당 기능도 현재 클래스에서 처리하는게 어색하다고 판단하여 수정하였습니다!

geunhwanlee commented on src/main/java/calculator/Split.java:
```diff
@@ -0,0 +1,43 @@
+      return new String[] {NULL_INPUT_VALUE};
+    }
+
+    return parseInput(input);
```
validateInput은 input을 검증하는 역할만을 하는것 같은데 문자열을 구분자로 분리하는 역할도 하고있네요.
validateInput가 검증 결과를 boolean으로 리턴해주는건 어떨까요?🤔

hyeongwon-up replied:
PR을 제출했을 때와 다시 보니까 많이 잘못된 점이 보이네요.

리뷰어님 말씀대로 validateInput이 분리하는 역할도 있는게 어색하네요.
parseInput에 validation의 결과(boolean) 에 의해서 로직을 진행하도록 수정을 해보았습니다.

geunhwanlee commented on src/main/java/calculator/Split.java:
```diff
@@ -0,0 +1,43 @@
+
+  private String[] parseInput(String input) {
+
+    Matcher matcher = Pattern.compile("//(.)\n(.*)").matcher(input);
```
상수로 선언한 CUSTOM_SPLIT_SEPARATOR_PATTERN을 사용해도 좋을것 같아요!☺️

geunhwanlee commented on src/main/java/calculator/Split.java:
```diff
@@ -0,0 +1,43 @@
+    Matcher matcher = Pattern.compile("//(.)\n(.*)").matcher(input);
+
+    if (matcher.find()) {
+      String customToken = matcher.group(1);
+      return matcher.group(2).split(customToken);
```
`matcher.group(1)` 에서 `1`이 의미하는바는 무엇일까요?
이런 숫자나 문자들을 매직넘버라고 하는데요
상수를 통해서 이름을 지정해 줄 수 있어요!😊

참고하실만한 링크 보내드릴게요
https://hoonmaro.tistory.com/44

hyeongwon-up replied:
'매직넘버' 라는 용어를 처음 들어보았고 구체적인 예시까지 첨부해주셔서 감사합니다.

matcher.group(1) 의 1은 은 커스텀 구분자가 위치해 있는 인덱스라는 표현을 하기위해 **CUSTOM_PATTERN_DELIMITER_INDEX** 라는 상수,
matcher.group(2) 는 커스텀 구분자가 들어가 있는 문자가 위치해 있는 인덱스라는 표현하기 위해  **CUSTOM_PATTERN_TEXT_INDEX** 상수로 표현해보았습니다.

geunhwanlee commented on src/main/java/calculator/StringAddCalculator.java:
```diff
@@ -0,0 +1,15 @@
+public class StringAddCalculator {
+
+  public static int splitAndSum(String input) {
+    Split split = new Split(input);
+
+    String[] stringArray = split.getStringArray();
+
+    StringSum stringSum = new StringSum(stringArray);
+
+    return stringSum.getSum();
```
Split과 StringSum은 input에 대한 output을 내는 유틸리티 클래스 성격이 강해보여요!

이런 경우엔 정적 메소드로 기능을 제공하는 것이 가독성과 사용성 측면에서 더 유리한 것 같아요.

```suggestion
    String[] stringArray = Split.getStringArray(input);

    return StringSum.getSum(stringArray);
```

hyeongwon-up replied:
정말 그러네요. 
뭔가 자바를 자바스럽게 아직 사용하는게 많이 미숙한 것 같아, 더 노력해야겠습니다.
말씀하신대로 인스턴스를 만들면서 사용하는 모습에 비하여  정적으로 사용하니 가독성과 사용성측면에서 훨씬 좋아보이네요.
수정하였습니다!.

geunhwanlee commented on src/main/java/calculator/Split.java:
```diff
@@ -0,0 +1,43 @@
+
+  private static final String NULL_INPUT_VALUE = "0";
+  private static final String DEFAULT_SPLIT_SEPARATORS= ",|:";
+  private static final Pattern CUSTOM_SPLIT_SEPARATOR_PATTERN = Pattern.compile("//(.)\n(.*)");
```
Pattern을 상수로 선언하신것 너무 좋습니다.👍🏻

geunhwanlee commented on src/main/java/calculator/StringSum.java:
```diff
@@ -0,0 +1,52 @@
+
+      return intValue;
+    } catch (NumberFormatException e) {
+      throw new RuntimeException(ERROR_MESSAGE_BY_NOT_INT_VALUE);
```
RuntimeException보다는 더 구체적인 예외를 던지는것이 좋을것 같아요!

자바에선 RuntimeException을 상속하는 여러 예외들을 제공하고있는데요
어떤 예외를 던지는게 좋을까요?🤔

hyeongwon-up replied:
넵! 제 생각에는 잘못된 입력을 받아서 발생해야하는 예외이니까.
**illegalargumentexception** 가 적합해 보여서 해당 예외로 처리하도록 수정을 해보았습니다.

geunhwanlee reviewed (changes requested):
안녕하세요 형원님!

2단계 미션 너무 잘해주셨어요!👏

몇가지 간단한 코멘트를 드렸는데요
확인 해 보시고 수정 후 리뷰요청 주세요.☺️

hyeongwon-up committed:
refactor(Split.java): 비어있는 입력값이 들어왔을 경우 비어있는 문자열 배열이 나오도록 수정.

- test(SplitTest.java): 비어있는 입력값이 들어왔을 경우, 비어있는 문자열이 나오는지 확인하도록 수정.

hyeongwon-up committed:
refactor(Split.java):

- input에 대한 validation 에 대한 기능을 분리하여 한 가지 기능만 담당하도록 수정.
- 상수로 선언해놓은 CUSTOM_SPLIT_SEPARATOR_PATTERN 을 사용하도록 수정.
- 매직 넘버를 사용하지 않고 상수로 선언하여 사용하도록 수정.

hyeongwon-up committed:
refactor(Split.java, StringSum.java, StringAddCalculator.java):

- Split.java, StringSum.java: Split 과 StringSum 을 유틸리티성으로 정적메소드로 제공하여 가독성 및 사용성 측면을 좋게 수정.
- StringAddCalculator.java: 수정된 정적메소드를 사용하도록 수정.
- SplitTest.java, StringSumTest.java: 수정된 정적메소드로 테스트 하도록 수정.

hyeongwon-up committed:
refactor(StringSum.java): 잘못된 input에 대해서 구체적인 예외를 발생시키도록 수정.

geunhwanlee commented on src/main/java/calculator/Split.java:
```diff
@@ -0,0 +1,42 @@
+
+  private static  boolean isValidInput(String input) {
+
+    if(input == null || input.isEmpty()) {
+      return false;
+    }
+
+    return true;
```
```suggestion
    return input != null && !input.isEmpty();
```
이렇게도 해볼 수 있겠네요!

geunhwanlee reviewed (approved):
미션 2단계 잘 해주셨습니다!👏

이번 단계는 여기서 머지하도록 할게요.

그럼 다음 단계 화이팅하세요!🔥
